### ISSUE TITLE
How to ensure reliability of custom user environments

### ISSUE THREAD

**Issue by choldgraf** [2i2c-org/infrastructure#1016]
### Background and proposal

**Context**

For most communities, we [recommend they use this image template](https://github.com/2i2c-org/hub-user-image-template) to define their environment via the repo2docker action.
For other communities, they'll want an environment that is complex and/or custom, and that they'd like to control entirely on their own. In those cases, we recommend that they define their own `Dockerfile` with their full user image environment.

There are two challenges with this setup:

**Potential incompatibility issues**

By letting communities define their own environment, we run the risk that they install packages that are incompatible with 2i2c's hub infrastructure. For example, we may need to enforce version matches for the `jupyterhub` package, or for `dask-gateway`.

**Complex environments that are hard to debug and break with our infrastructure**

For particularly complex environments, we run the risk that the environment is so complex that if something breaks, it will be very difficult for a 2i2c engineer to have the context needed to debug. An example of this are [the Pangeo docker images](https://github.com/pangeo-data/pangeo-docker-images).

An example of this situation recently came up when @rabernat reported that Dask Gateway was not functioning after Pangeo released a new Docker image and updated it in their configurator. Here's the diff between the old and new image:
   
https://github.com/pangeo-data/pangeo-docker-images/compare/2021.10.19...2022.02.04

There's a lot of stuff that is changed in there, and I think it'll be really hard for a person who isn't deeply familiar with that image to debug efficiently.

We need some way to give communities the flexibility to control their own environment, but provide guard-rails and QA that makes it less-likely they'll change the environment in a way that creates unexpected breakages. 

### Implementation guide and constraints

There are a few potential ways that we could improve this issue, I'd love feedback about any of them:

#### Define an environment specification and validate it

We could define a specification for the environment that creates certain conditions that must be met (e.g. jupyterhub version XXX, dask-gateway version YYY, configuration value ZZZ).

We could then define a GitHub Action that runs with our CI/CD and runs some basic validation of the user image environment to ensure certain conditions are met. 

#### Testing automation

Make it easy for a community to define their own continuous integration test that runs as a part of any new deploys for that hub. For example, we could use `nbclient` to run a notebook and ensure it passes as a part of our CI/CD.

#### Encourage communities to build off of the 2i2c Dockerfile

2i2c maintains a Dockerfile as a base for its hubs here: https://github.com/2i2c-org/2i2c-hubs-image/blob/main/Dockerfile

We could encourage communities to use this Dockerfile as a base, and only add necessary packages on top of it. This could give us more control over the major infrastructure pieces (JupyterHub, Dask Gateway, etc) that need to work across all hubs.

**However**, it would also put 2i2c directly on the dependency chain, which could cause many headaches for both us and our communities.

**Comment by choldgraf** [comment]
I'd love feedback from @rabernat since he was the one most recently bitten by this one :-)

**Comment by yuvipanda** [comment]
Paging @jacobtomlinson too as it looks like he's working on a fix for this https://twitter.com/_JacobTomlinson/status/1496027340505243648

**Comment by jacobtomlinson** [comment]
Yep I'm working on a generic tool for validating containers like this. 

Often I find that platforms allow you to bring your own container but only list a set of requirements in their docs for things like installed packages, non-root, home directory, etc. I'm building a little tool to codify these requirements and validate containers against them.

My main goal is adding to the CI for the Dask and RAPIDS containers to test that they are compatible with KubeFlow, Binder, JupyterHub, Dask Cloud Provider, Dask Kubernetes, Dask Gateway, etc. Supporting so many things means regressions are happening more frequently.

I'm nearly at the point of having a POC written. I just need to see if I need to run it through the NVIDIA open sourcing process which can take a little while.

**Comment by rabernat** [comment]
Here is my main feedback about how to approach this: integration testing via a gallery of example worflows.

In fact, we already have such a gallery: http://gallery.pangeo.io/ (or more specifically for my use cases, http://gallery.pangeo.io/repos/pangeo-gallery/physical-oceanography/)

Running through these notebook in this gallery (just a github repo under the hood), either manually or automatically, would immediately catch any problems that would arise after changes to the image or hub configuration. That would be my best recommendation in terms of general practices for 2i2c.

**Comment by jacobtomlinson** [comment]
Yeah that would be awesome in addition to the unit testing approach I'm taking.

**Comment by choldgraf** [comment]
@rabernat totally agree, I think that the MVP version of this might just be a specification for the user environment. For example:

- As a part of testing any changes to a hub in CI/CD, do this:
- If the user environment defines a folder called `tests` (or maybe `.2i2c/tests`?), then:
- Loop through all files in the folder
- For any `ipynb` files (or `jupytext` notebooks), execute them top-to-bottom somehow and fail if there are any errors

Over time we could add extra things like a place to look at the notebook that failed etc, but this could be a relatively easy start.

**Comment by rabernat** [comment]
The unit testing sounds super useful! But unit testing would never have caught https://github.com/2i2c-org/infrastructure/issues/1031 because it involves interaction between the container and other services.

**Comment by rabernat** [comment]
@choldgraf FWIW, binderbot already does items 3 and 4 on your list in the context of a repo + a binder spec.

**Comment by damianavila** [comment]
I think a mix of:

1. Hierarchical set of images
2. Unit test 
3. Integration test
will get us there in a majority of the cases.

1 and 2 are probably easier than 3, although 3 has a bigger validation power, IMHO.

**Comment by jacobtomlinson** [comment]
The container validation tool I built is now open source! I'm going to start using it in CI everywhere I am building containers for platforms like kubeflow, dask-gateway, jupyterhub, etc.

https://github.com/NVIDIA/container-canary

It's early and buggy, but feedback and contributions are very welcome!

**Comment by yuvipanda** [comment]
Awesome, @jacobtomlinson! We should integrate this into https://github.com/jupyterhub/repo2docker-action as a best practice.
